drivers/usb/uhc/uhc_stm32.c

```
@@ -0,0 +1,1594 @@
+			     struct net_buf *const buf, const uint8_t ep_type)
+{
+	size_t len = net_buf_tailroom(buf);
+	void *buffer_tail = net_buf_add(buf, len);
+
+	int err = priv_request_submit(priv, chan_num, 1, ep_type, 1, buffer_tail, len);
+
+	if (err) {
+		net_buf_remove_mem(buf, len);
+		return err;
+	}
+
+	return 0;
+}
+
+static int priv_ongoing_xfer_control_run(struct uhc_stm32_data *const priv,
+					 struct uhc_transfer *const xfer)
+{
+	if (xfer->stage == UHC_CONTROL_STAGE_SETUP) {
+		return priv_control_setup_send(priv,
+			priv->ongoing_xfer_pipe_id, xfer->setup_pkt, sizeof(xfer->setup_pkt)
+		);
+	}
+
+	if (xfer->buf != NULL && xfer->stage == UHC_CONTROL_STAGE_DATA) {
+		if (USB_EP_DIR_IS_IN(xfer->ep)) {
+			return priv_data_receive(priv,
+				priv->ongoing_xfer_pipe_id, xfer->buf, USB_EP_TYPE_CONTROL
+			);
+		} else {
+			return priv_data_send(priv,
+				priv->ongoing_xfer_pipe_id, xfer->buf,
+				USB_EP_TYPE_CONTROL, xfer->mps
+			);
+		}
+	}
+
+	if (xfer->stage == UHC_CONTROL_STAGE_STATUS) {
+		if (USB_EP_DIR_IS_IN(xfer->ep)) {
+			return priv_control_status_send(priv, priv->ongoing_xfer_pipe_id);
+		} else {
+			return priv_control_status_receive(priv, priv->ongoing_xfer_pipe_id);
+		}
+	}
+
+	return -EINVAL;
+}
+
+static int priv_ongoing_xfer_bulk_run(struct uhc_stm32_data *const priv,
+				      struct uhc_transfer *const xfer)
+{
+	if (USB_EP_DIR_IS_IN(xfer->ep)) {
+		return priv_data_receive(priv,
+			priv->ongoing_xfer_pipe_id, xfer->buf, USB_EP_TYPE_BULK
+		);
+	} else {
+		return priv_data_send(priv,
+			priv->ongoing_xfer_pipe_id, xfer->buf, USB_EP_TYPE_BULK, xfer->mps
+		);
+	}
+}
+
+static int priv_ongoing_xfer_run(struct uhc_stm32_data *const priv)
+{
+	if (USB_EP_GET_IDX(priv->ongoing_xfer->ep) == USB_CONTROL_ENDPOINT) {
+		return priv_ongoing_xfer_control_run(priv, priv->ongoing_xfer);
+	}
+
+	/* TODO: For now all other xfer are considered as bulk transfers,
+	 * add support for all type of transfer.
```

can you return an error in such non-bulk case? to prevent hidden issue.